docs/site/Loopback-component-authentication.md

```diff
@@ -0,0 +1,596 @@
+
+The `AuthenticationStrategyProvider` class **(shown below)** declares an
+`extension point` named
+`AuthenticationBindings.AUTHENTICATION_STRATEGY_EXTENSION_POINT_NAME` via the
+`@extensionPoint` decorator. The binding scope is set to **transient** because
+an authentication strategy **may** differ with each request. With the aid of the
+`@extensions()` **getter** decorator, `AuthenticationStrategyProvider` is
+responsible for **finding** and **returning** an authentication strategy which
+has a specific **name** and has been `registered` as an **extension** of the
+aforementioned **extension point**.
+
+```ts
+@extensionPoint(
+  AuthenticationBindings.AUTHENTICATION_STRATEGY_EXTENSION_POINT_NAME,
+  {scope: BindingScope.TRANSIENT},
+) //this needs to be transient, e.g. for request level context.
+export class AuthenticationStrategyProvider
+  implements Provider<AuthenticationStrategy | undefined> {
+  constructor(
+    @extensions()
+    private authenticationStrategies: Getter<AuthenticationStrategy[]>,
+    @inject(AuthenticationBindings.METADATA)
+    private metadata?: AuthenticationMetadata,
+  ) {}
+  async value(): Promise<AuthenticationStrategy | undefined> {
+    if (!this.metadata) {
+      return undefined;
+    }
+    const name = this.metadata.strategy;
+    const strategy = await this.findAuthenticationStrategy(name);
+    if (!strategy) {
+      // important to throw a non-protocol-specific error here
+      let error = new Error(`The strategy '${name}' is not available.`);
+      Object.assign(error, {
+        code: AUTHENTICATION_STRATEGY_NOT_FOUND,
+      });
+      throw error;
+    }
+    return strategy;
+  }
+
+  async findAuthenticationStrategy(name: string) {
+    const strategies = await this.authenticationStrategies();
+    const matchingAuthStrategy = strategies.find(a => a.name === name);
+    return matchingAuthStrategy;
+  }
+}
+```
+
+{% include note.html content="
+The specific <b>name</b> that it looks for is a unique name you specify as a parameter in an <b>authentication decorator</b> applied to a controller endpoint. This decorator is discussed in the next section.
+" %}
+
+In order for your custom authentication strategy to be **found**, it needs to be
+**registered**.
+
+**Registering** a custom authentication strategy `BasicAuthenticationStrategy`
+as an extension of the
+`AuthenticationBindings.AUTHENTICATION_STRATEGY_EXTENSION_POINT_NAME` extension
+point in an application `application.ts` is as simple as:
+
+```ts
+registerAuthenticationStrategy(this, BasicAuthenticationStrategy);
+```
+
+## Authentication Decorator
+
+Once you have a **registered** authentication strategy with a **unique** name,
+then you are ready to use the
+[Authentication Decorator](https://loopback.io/doc/en/lb4/Decorators_authenticate.html)
+on your controller endpoints.
+
+The decorator's syntax is :
+
+```ts
+@authenticate(strategyName: string, options?: Object)
```

Comment:
@jannyHou  thanks for your feedback.

> A more accurate type would be Promise<UserProfile> 

When creating this diagram, space is very limited, so I chose to avoid using stuff like:
`this.` and `Promise<  >` to save space, and give a general idea of the logic.  The documentation text in the sections below the diagram will show the correct code.

> nitpick: 'some' --> 'amodel' or 'somemodel'. Just a suggestion based on personally preferred naming conversion. 

The `lb4 controller` CLI creates controller files is this format: someName.controller.ts . And all the artifacts follow the same pattern. I think my filename is ok.


> Maybe re-organize the rectangles in a more sorted way? 

It took a long time to create the diagram and all its pieces and labels, so I am reluctant iteratively move all the PowerPoint shapes/labels/arrows around  `unless` someone sends me a `pencil and paper picture` of a re-design.

Reply:
> When creating this diagram, space is very limited, so I chose to avoid using stuff like:
this. and Promise< > to save space

👍 Fair enough.

And the naming and diagram comments are just nitpicks 🙂 feel free to ignore.